### PR TITLE
gh-124872: Replace enter/exit events with "switched"

### PR DESCRIPTION
**Pull request by Eclips4**
* Issue: gh-125512

----
📚 Documentation preview 📚: https://cpython-previews--125532.org.readthedocs.build/

* Issue: gh-124872

### CODE REVIEW

**Comment by Eclips4** [comment]
With that PR, `test_capi` is no longer leaking:
```python
admin@Admins-MacBook-Air ~/p/cpython (revert-of-revert)> ./python.exe -m test -R 3:3 test_capi
Using random seed: 992095454
0:00:00 load avg: 26.10 Run 1 test sequentially in a single process
0:00:00 load avg: 26.10 [1/1] test_capi
beginning 6 repetitions. Showing number of leaks (. for 0 or less, X for 10 or more)
123:456
XX. ...
test_capi passed in 43.3 sec

== Tests result: SUCCESS ==

1 test OK.

Total duration: 43.4 sec
Total tests: run=985 skipped=67
Total test files: run=1/1
Result: SUCCESS
```

**Comment by vstinner** [comment]
>  gh-125512: Revert #125513 and fix reference cycle #125532 

It's not a revert, but a revert of a revert :-D The PR title should be `gh-124872: Replace enter/exit events with "switched`.

**Comment by vstinner** [comment]
I suggest to wait after Python 3.14.0 alpha1 release to retry this change.

**Comment by 1st1** [comment]
> I suggest to wait after Python 3.14.0 alpha1 release to retry this change.

Why? We can debug and fix the ref leak. 

@rhansen Please build CPython with `./configure --with-pydebug` and run the tests with `-R3:3`. That will take much longer so I suggest just running asyncio & contextvars & decimal tests with this flag.

**Comment by vstinner** [comment]
me:
> I suggest to wait after Python 3.14.0 alpha1 release to retry this change.

@1st1:
> Why?

It's just that there are many broken buildbots, I think that this change can wait for 3.14.0a2.

**Comment by Eclips4** [comment]
FYI, I ran the test suite in hunt refleaks mode locally and it succeeded:
```python
== Tests result: SUCCESS ==

25 tests skipped:
    test.test_asyncio.test_windows_events
    test.test_asyncio.test_windows_utils test.test_gdb.test_backtrace
    test.test_gdb.test_cfunction test.test_gdb.test_cfunction_full
    test.test_gdb.test_misc test.test_gdb.test_pretty_print
    test.test_multiprocessing_fork.test_manager
    test.test_multiprocessing_fork.test_misc
    test.test_multiprocessing_fork.test_processes
    test.test_multiprocessing_fork.test_threads test_android
    test_dbm_gnu test_devpoll test_epoll test_free_threading
    test_launcher test_msvcrt test_perf_profiler test_perfmaps
    test_startfile test_winapi test_winconsoleio test_winreg test_wmi

11 tests skipped (resource denied):
    test_curses test_peg_generator test_pyrepl test_smtpnet
    test_socketserver test_tkinter test_ttk test_urllib2net
    test_urllibnet test_winsound test_zipfile64

442 tests OK.

Total duration: 20 min 5 sec
Total tests: run=44,347 skipped=2,182
Total test files: run=467/478 skipped=25 resource_denied=11
Result: SUCCESS
```

**Comment by 1st1** [comment]
@Eclips4 thanks for fixing the leak!

@vstinner 
>It's just that there are [many broken buildbots](https://buildbot.python.org/#/wsgi/release_status), I think that this change can wait for 3.14.0a2.

I sympathize, really sorry about breaking them and causing pain :/ Can we try landing this PR since it appears it should solve the issue? I'm afraid if we wait on this for a month to get it merged it will go stale and we just forget. I'd try again and if it fails we wait? If however you really don't want to do it, then fine.

**Comment by ambv** [comment]
Since we're about to release a1 tonight, I'd wait with landing this until right after a1. We won't forget about it.

**Comment by ambv** [comment]
Actually, @hugovk should make the call.

File: Lib/test/test_capi/test_watchers.py [hunk 1]
```diff
@@ -577,68 +577,66 @@ class TestContextObjectWatchers(unittest.TestCase):
...
+                ctx_outer.run(_in_outer)
+                self.assertEqual([x.err_msg for x in unraisables],
+                                ["Exception ignored in Py_CONTEXT_SWITCHED "
+                                f"watcher callback for {ctx!r}"
+                                for ctx in [ctx_inner, ctx_outer]])
+                self.assertEqual([str(x.exc_value) for x in unraisables],
+                                ["boom!", "boom!"])
+            finally:
+                # Break reference cycle
+                unraisables = None
```
Comment (rhansen) [inline Lib/test/test_capi/test_watchers.py]:
```suggestion
        try:
            ctx_outer.run(_in_outer)
            self.assertEqual([x.err_msg for x in unraisables],
                             ["Exception ignored in Py_CONTEXT_SWITCHED "
                              f"watcher callback for {ctx!r}"
                              for ctx in [ctx_inner, ctx_outer]])
            self.assertEqual([str(x.exc_value) for x in unraisables],
                             ["boom!", "boom!"])
        finally:
            # Break reference cycle
            unraisables = None
```

I tested this change by running:

```shell
make test TESTOPTS="-R3:3 test.test_capi.test_watchers test.test_context test.test_asyncio test.test_decimal" TESTTIMEOUT=0
```

**Review by rhansen — CHANGES_REQUESTED** [review]
Apologies for the breakage, and thank you for debugging!

Is a blurb required now that there's an API change between v3.14.0a1 and v3.14.0a2?

Also, would you mind fixing up the commit message, authorship info, etc. on the first commit:

```shell
git -c sequence.editor='sed -i -e "1s/^pick/edit/"' rebase -i main &&
git commit --amend -C 843d28f59d2 &&
git rebase --continue
```

**Comment by 1st1** [comment]
@ambv @vstinner 

> Since we're about to release a1 tonight, I'd wait with landing this until right after a1. We won't forget about it.

Oh I was reading the schedule wrong. I didn't realize Victor wants to delay it just one day, essentially, I thought we were talking one month to merge this. Absolutely can wait a few days.

**Comment by hugovk** [comment]
Thanks, a1 is now out, merge whenever you're ready.

**Comment by vstinner** [comment]
I applied @rhansen's change, before that, the test did nothing :-)

> Also, would you mind fixing up the commit message, authorship info, etc. on the first commit:

That sounds like a good idea :-) Or at least, make sure that @rhansen is mentioned as a co-author of the change.

**Comment by Eclips4** [comment]
> Apologies for the breakage, and thank you for debugging!
> 
> Is a blurb required now that there's an API change between v3.14.0a1 and v3.14.0a2?
> 
> Also, would you mind fixing up the commit message, authorship info, etc. on the first commit:
> 
> ```shell
> git -c sequence.editor='sed -i -e "1s/^pick/edit/"' rebase -i main &&
> git commit --amend -C 843d28f59d2 &&
> git rebase --continue
> ```

Sorry about that! I definitely will do it.

**Comment by Eclips4** [comment]
@rhansen it seems I have done it. If you agree, I will merge it.

**Comment by vstinner** [comment]
> (I'm actually do not understand why for f0b3aa862b629690496d0336904a9bb1ada6dd47 github is treating me as co-author...)

GitHub is correct when saying:

> rhansen **authored** and Eclips4 **committed**

**Review by vstinner — APPROVED** [review]
LGTM

**Comment by ZeroIntensity** [comment]
Should this get tested with the refleak buildbots?

**Comment by vstinner** [comment]
> Should this get tested with the refleak buildbots?

I don't think so. I tested test_capi manually, and @Eclips4 [ran the whole test suite](https://github.com/python/cpython/pull/125532#issuecomment-2414275256) with `-R 3:3`.

**Pull request merged**